### PR TITLE
CuPy-Xarray post

### PR DESCRIPTION
**Pull request by negin513**
First draft of a post about [CuPy-Xarray updated docs ](https://cupy-xarray.readthedocs.io/).

### CODE REVIEW

File: src/posts/cupy-tutorial/index.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+title: 'CuPy-Xarray: Xarray on GPUs!'
+date: '2023-10-27'
+authors:
+  - name: Negin Sobhani, Deepak Cherian, Max Jones
+    github: negin513, dcherian, maxrjones
```
Comment (andersy005) [inline src/posts/cupy-tutorial/index.md:6]:
```suggestion
  - name: Negin Sobhani
    github: negin513
 - name: Deepak Cherian
    github: dcherian
 - name: Max Jones
    github: maxrjones
```

File: src/posts/cupy-tutorial/index.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+```shell
+pip install cupy-xarray
+```
+
+## Appendix II: Additional Resources
+
+[CuPy User Guide](https://docs.cupy.dev/en/stable/user_guide/index.html)
+[Xarray User Guide](https://docs.xarray.dev/en/stable/user-guide/index.html)
+[Cupy-Xarray Github](https://github.com/xarray-contrib/cupy-xarray.git)
```
Comment (weiji14) [inline src/posts/cupy-tutorial/index.md:69]: Add your SciPy 2023 tutorial at https://negin513.github.io/cupy-xarray-tutorials?
```suggestion
[Cupy-Xarray Github](https://github.com/xarray-contrib/cupy-xarray.git)
[Xarray with GPUs tutorial](https://negin513.github.io/cupy-xarray-tutorials)
```

File: src/posts/cupy-tutorial/index.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+## TLDR
+
+The [CuPy-Xarray](https://github.com/xarray-contrib/cupy-xarray) collaboration brings GPU acceleration into Xarray workflows. Explore the new documentation and tutorials to leverage these capabilities for your data science needs. 🎉 🥳 🚀
+
+## Background
+
+### What is CuPy-Xarray?
+
+CuPy is a GPU-accelerated library for numerical computations. It’s like NumPy but runs on NVIDIA CUDA. Meanwhile, Xarray is an open-source Python library that simplifies working with labelled multi-dimensional arrays, such as in weather data, satellite images, and more. Xarray can wrap custom duck array objects (i.e. NumPy-like arrays) that follow specific protocols.
```
Comment (weiji14) [inline src/posts/cupy-tutorial/index.md:18]: Add a link to CuPy - https://cupy.dev.
```suggestion
[CuPy](https://cupy.dev) is a GPU-accelerated library for numerical computations. It’s like NumPy but runs on NVIDIA CUDA devices. Meanwhile, Xarray is an open-source Python library that simplifies working with labelled multi-dimensional arrays, such as in weather data, satellite images, and more. Xarray can wrap custom duck array objects (i.e. NumPy-like arrays) that follow specific protocols.
```

File: src/posts/cupy-tutorial/index.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+The new documentation offer the following sections:
+
+1. Basics of CuPy: This section provides an introductions to CuPy and basics of GPU computing and data transfer between host and device.
+2. Introduction to CuPy-Xarray : This section introduces the CuPy-Xarray library and how it handles Xarray DataArrays on GPUs.
+3. Basic Computations with CuPy-Xarray: This section introduces, how to use CuPy-Xarray to perform basic computations on Xarray DataArrays. Next, it introduces the concepts of broadcasting and alignment.
+4. High-level Computation with CuPy-Xarray: This section introduces, Applying high-level functions like groupby, resample, rolling, and apply_ufunc to xarray DataArray.
+5. Custom Kernels with `apply_ufunc` : Here we will talk about how to write custom kernels for `apply_ufunc` and how to use `apply_ufunc` with `groupby` and `resample`.
+6. A real world example: This section introduces, how to use CuPy-Xarray to accelerate a real world example. We will use the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to accelerate a real world example.
```
Comment (weiji14) [inline src/posts/cupy-tutorial/index.md:39]:
```suggestion
6. A real world example: This section introduces, how to use CuPy-Xarray to accelerate a real world example. We will use the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to speed-up computation on climate data variables.
```

File: src/posts/cupy-tutorial/index.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+If you have any questions, encounter issues, or want to contribute, the community forums are a great place to start.
```
Comment (weiji14) [inline src/posts/cupy-tutorial/index.md:41]: Link to the forum at https://discourse.pangeo.io?
```suggestion
If you have any questions, encounter issues, or want to contribute, the [community forum](https://discourse.pangeo.io) is a great place to start.
```

**Review by weiji14 — COMMENTED** [review]
A few more minor suggestions. This looks about ready to publish @negin513, were you planning to add anything else?

File: src/posts/cupy-tutorial/index.md (hunk 2 above)
Comment (negin513) [reply src/posts/cupy-tutorial/index.md:73]: @weiji14 I completely forgot about it. 😆  Thanks for your suggestions. Added! 

image

**Comment by negin513** [comment]
@andersy005 Can you please authorize the vercel build on this? Thanks. 🙏

File: src/posts/cupy-tutorial/index.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
+---
+title: 'CuPy-Xarray: Xarray on GPUs!'
+date: '2023-10-27'
+authors:
+  - name: Negin Sobhani
+    github: negin513
+ - name: Deepak Cherian
+   github: dcherian
+ - name: Max Jones
+   github: maxrjones
```
Comment (andersy005) [inline src/posts/cupy-tutorial/index.md:10]:
```suggestion
  - name: Negin Sobhani
    github: negin513
  - name: Deepak Cherian
    github: dcherian
  - name: Max Jones
    github: maxrjones
```

**Comment by negin513** [comment]
@dcherian and @maxrjones I wrote this initial draft for a blog post on the updated cupy-xarray docs. 
Please review this and let me know your comments.

**Comment by negin513** [comment]
@dcherian Should we add the funding source to this? Maybe in the acknowledgement section?

File: src/posts/cupy-tutorial/index.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+> CuPy-Xarray provides an interface for using CuPy in Xarray, providing [accessors](https://docs.xarray.dev/en/stable/internals/extending-xarray.html) on the Xarray objects.
+
+### Why is this important?
+
+GPU acceleration is becoming increasingly important in scientific research, data analysis, and AI/ML techniques due to its ability to perform massively parallel computations. GPUs can greatly accelerate the processing of array datasets, allowing for faster analysis and modeling of large datasets. By leveraging the power of GPUs with tools such as CuPy and CuPy-Xarray, Xarray users can gain significant performance improvements and unlock new opportunities for scientific discovery.
+
+## New Documentation and Tutorials
+
+We have recently created detailed documentation with examples to help users get started with CuPy-Xarray. Check it out at [this link](https://cupy-xarray.readthedocs.io/en/latest/.
```
Comment (maxrjones) [inline src/posts/cupy-tutorial/index.md:34]:
```suggestion
We have recently created detailed documentation with examples to help users get started with CuPy-Xarray. Check it out at [this link](https://cupy-xarray.readthedocs.io/en/latest/).
```

File: src/posts/cupy-tutorial/index.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+The new documentation offer the following topics:
+
+1. **[Basics of CuPy](https://cupy-xarray.readthedocs.io/source/cupy-basics.html)** : This section provides an introductions to CuPy and basics of GPU computing and data transfer between host and device.
+2. **[Introduction to CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/introduction.html)** : This section introduces the CuPy-Xarray library and how it handles Xarray DataArrays on GPUs.
+3. **[Basic Computations with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html)** : This section introduces, how to use CuPy-Xarray to perform basic computations on Xarray DataArrays. Next, it introduces the concepts of broadcasting and alignment.
```
Comment (maxrjones) [inline src/posts/cupy-tutorial/index.md:40]:
```suggestion
3. **[Basic Computations with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html)** : This section introduces how to use CuPy-Xarray to perform basic computations on Xarray DataArrays. Next, it introduces the concepts of broadcasting and alignment.
```

File: src/posts/cupy-tutorial/index.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+4. **[High-level Computation with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html#)** : This section introduces, Applying high-level functions like groupby, resample, rolling, and apply_ufunc to xarray DataArray.
```
Comment (maxrjones) [inline src/posts/cupy-tutorial/index.md:41]:
```suggestion
4. **[High-level Computation with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html#)** : This section introduces applying high-level functions like groupby, resample, rolling, and apply_ufunc to xarray DataArray.
```

File: src/posts/cupy-tutorial/index.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+5. **[Custom Kernels with `apply_ufunc`](https://cupy-xarray.readthedocs.io/source/apply-ufunc.html)** : Here we will talk about how to write custom kernels for `apply_ufunc` and how to use `apply_ufunc` with `groupby` and `resample`.
+6. **[A real world example](https://cupy-xarray.readthedocs.io/source/real-example-1.html)** : This section introduces, how to use CuPy-Xarray to accelerate a real world earth system model analysis workflow. In this demo, we used the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to speed-up computations on climate data variables.
```
Comment (maxrjones) [inline src/posts/cupy-tutorial/index.md:43]:
```suggestion
6. **[A real world example](https://cupy-xarray.readthedocs.io/source/real-example-1.html)** : This section introduces how to use CuPy-Xarray to accelerate a real world earth system model analysis workflow. In this demo, we used the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to speed-up computations on climate data variables.
```

**Review by maxrjones — COMMENTED** [review]
Nice post!

File: src/posts/cupy-tutorial/index.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  - name: Deepak Cherian
+    github: dcherian
+  - name: Max Jones
+    github: maxrjones
+summary: 'CuPy-Xarray is a Python library that leverages CuPy, a GPU array library, and Xarray, a library for multi-dimensional labeled array computations, to enable fast and efficient data processing on GPUs.'
+---
+
+## TLDR
+
+The [CuPy-Xarray](https://github.com/xarray-contrib/cupy-xarray) collaboration brings GPU acceleration into Xarray workflows. Explore the [new documentation](https://cupy-xarray.readthedocs.io/) and tutorials to explore how CuPy-Xarray enables GPU accelerations on large multidimensional datasets. 🎉 🥳 🚀
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:16]:
```suggestion
The [CuPy-Xarray](https://github.com/xarray-contrib/cupy-xarray) project makes mixing GPU acceleration with Xarray workflows very convenient! Explore the [new documentation](https://cupy-xarray.readthedocs.io/) and tutorials to explore how CuPy-Xarray enables GPU accelerations on large multidimensional datasets. 🎉 🥳 🚀
```

File: src/posts/cupy-tutorial/index.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+## Background
+
+### What is CuPy-Xarray?
+
+[CuPy](https://cupy.dev) is a GPU-accelerated library for numerical computations. It’s like NumPy but runs on NVIDIA CUDA devices. Meanwhile, Xarray is an open-source Python library that simplifies working with labelled multi-dimensional arrays, such as in weather data, satellite images, and more. Xarray can wrap custom duck array objects (i.e. NumPy-like arrays) that follow specific protocols.
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:22]:
```suggestion
[CuPy](https://cupy.dev) is a GPU-accelerated library for numerical computations. CuPy provides a NumPy-like array object -- a duck array -- that follows various standard array protocols and executes computations on CUDA-capable devices. Xarray can [wrap duck array](https://docs.xarray.dev/en/stable/user-guide/duckarrays.html) objects (i.e. NumPy-like arrays) that follow specific protocols.
```

File: src/posts/cupy-tutorial/index.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+The CuPy-Xarray integration means that Xarray can now handle CuPy arrays, providing a seamless transition between CPU and GPU computations in your data pipeline.
+
+> CuPy-Xarray provides an interface for using CuPy in Xarray, providing [accessors](https://docs.xarray.dev/en/stable/internals/extending-xarray.html) on the Xarray objects.
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:26]:
```suggestion
Thus Xarray can handle CuPy arrays, and `cupy-xarray` provides a number of useful methods under the `xarray_object.cupy` namespace, allowing seamless transition between CPU and GPU computations in your data pipeline.
```

File: src/posts/cupy-tutorial/index.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+3. **[Basic Computations with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html)** : This section introduces how to use CuPy-Xarray to perform basic computations on Xarray DataArrays. Next, it introduces the concepts of broadcasting and alignment.
+4. **[High-level Computation with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html#)** : This section introduces applying high-level functions like groupby, resample, rolling, and apply_ufunc to xarray DataArray.
+5. **[Custom Kernels with `apply_ufunc`](https://cupy-xarray.readthedocs.io/source/apply-ufunc.html)** : Here we will talk about how to write custom kernels for `apply_ufunc` and how to use `apply_ufunc` with `groupby` and `resample`.
+6. **[A real world example](https://cupy-xarray.readthedocs.io/source/real-example-1.html)** : This section introduces how to use CuPy-Xarray to accelerate a real world earth system model analysis workflow. In this demo, we used the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to speed-up computations on climate data variables.
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:43]:
```suggestion
1. **[Basics of CuPy](https://cupy-xarray.readthedocs.io/source/cupy-basics.html)** : An introduction to CuPy, basics of GPU computing, and data transfer between host and device.
2. **[Introduction to CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/introduction.html)** 
3. **[Basic Computations with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html)** 
4. **[High-level Computation with CuPy-Xarray](https://cupy-xarray.readthedocs.io/source/basic-computations.html#)** :Applying high-level functions like groupby, resample, rolling, and apply_ufunc to xarray DataArray.
5. **[Custom Kernels with `apply_ufunc`](https://cupy-xarray.readthedocs.io/source/apply-ufunc.html)** : Custom CUDA kernels for `apply_ufunc` and how to use `apply_ufunc` with `groupby` and `resample`.
6. **[A real world example](https://cupy-xarray.readthedocs.io/source/real-example-1.html)** : This section introduces how to use CuPy-Xarray to accelerate a real world earth system model analysis workflow. In this demo, we used the NASA Earth Exchange Global Daily Downscaled Projections (NEX-GDDP-CMIP6) to demonstrate how to use CuPy-Xarray to speed-up computations on climate data variables.
```

File: src/posts/cupy-tutorial/index.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+If you have any questions, encounter issues, or want to contribute, the [community forum](https://discourse.pangeo.io) is a great place to start.
+
+## Summary
+
+CuPy-Xarray is a Python library that leverages CuPy, a GPU array library, and Xarray, a library for multi-dimensional labeled array computations, to enable fast and efficient data processing on GPUs. By combining the capabilities of CuPy and Xarray, CuPy-Xarray provides a convenient interface for performing accelerated computations and analysis on large multidimensional datasets. With the new documentation and tutorials, users can quickly adapt to this integration and optimize their data science workflows.🚀
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:49]:
```suggestion
CuPy-Xarray is a Python library helps you use CuPy, a GPU array library, and Xarray, a library for multi-dimensional labeled array computations, to enable fast and friendly data processing on GPUs. With the new documentation and tutorials, users can quickly adapt to this integration and optimize their data science workflows.🚀
```

File: src/posts/cupy-tutorial/index.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+The CuPy-Xarray integration means that Xarray can now handle CuPy arrays, providing a seamless transition between CPU and GPU computations in your data pipeline.
+
+> CuPy-Xarray provides an interface for using CuPy in Xarray, providing [accessors](https://docs.xarray.dev/en/stable/internals/extending-xarray.html) on the Xarray objects.
+
+### Why is this important?
+
+GPU acceleration is becoming increasingly important in scientific research, data analysis, and AI/ML techniques due to its ability to perform massively parallel computations. GPUs can greatly accelerate the processing of array datasets, allowing for faster analysis and modeling of large datasets. By leveraging the power of GPUs with tools such as CuPy and CuPy-Xarray, Xarray users can gain significant performance improvements and unlock new opportunities for scientific discovery.
+
```
Comment (dcherian) [inline src/posts/cupy-tutorial/index.md:31]: Talk about https://github.com/cupy/cupy/pull/7575 and https://github.com/cupy/cupy/pull/7811 too!

**Review by dcherian — COMMENTED** [review]
Great work, mostly minor comments, just one major suggestion to talk about your upstream cupy work!

**Comment by dcherian** [comment]
@negin513 thoughts on the suggestions?

File: src/posts/cupy-tutorial/index.md (hunk 16 above)
Comment (negin513) [reply src/posts/cupy-tutorial/index.md:29]: Done. Thanks for your great comments.

**Review by dcherian — APPROVED** [review]

**Comment by dcherian** [comment]
@andersy005 this is good to go. Let's merge! (or maybe take a quick read before merging :) )

File: src/posts/cupy-tutorial/index.md [hunk 17]
```diff
@@ -0,0 +1,77 @@
+---
+title: 'CuPy-Xarray: Xarray on GPUs!'
+date: '2023-10-27'
```
Comment (weiji14) [inline src/posts/cupy-tutorial/index.md:3]: Update date before merging :innocent:
```suggestion
date: '2023-11-12'
```

**Review by weiji14 — APPROVED** [review]

**Comment by maxrjones** [comment]
I did one last copy edit (including updating the date). @andersy005 would you be willing commit the last suggestions and merge this for us?

**Review by maxrjones — APPROVED** [review]

**Review by andersy005 — APPROVED** [review]
thank you, @negin513 / @dcherian / @maxrjones / @weiji14! So glad to see this brought to completion.

**Comment by dcherian** [comment]
https://twitter.com/xarray_dev/status/1748007050842431922

**Pull request merged**